firmware/bittide-sys/tests/program_stream.rs

```diff
@@ -0,0 +1,151 @@
+// SPDX-FileCopyrightText: 2022 Google LLC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    path::{Path, PathBuf},
+    process::Command,
+};
+
+use proptest::prelude::*;
+use test_strategy::proptest;
+
+use bittide_sys::program_stream::*;
+
+mod elf_common;
+use elf_common::*;
+
+fn find_cabal_root() -> Option<PathBuf> {
+    let current_path = std::env::current_dir().ok()?;
+
+    fn find_root(path: PathBuf) -> Option<PathBuf> {
+        let manifest = path.join("cabal.project");
+        if manifest.exists() {
+            return Some(path);
+        }
+        let parent = path.parent()?;
+
+        find_root(parent.to_owned())
+    }
+
+    find_root(current_path)
+}
+
+fn build_program_stream_exec(root_path: &Path) -> bool {
+    let status = Command::new("cabal")
+        .args(["build", "program-stream"])
+        .current_dir(root_path)
+        .status();
+
+    let Ok(status) = status else { return false };
+
+    status.success()
+}
+
+fn program_stream_exec_path() -> Option<PathBuf> {
+    let root = find_cabal_root()?;
+    let compile_success = build_program_stream_exec(&root);
+
+    if !compile_success {
+        eprintln!("building of program-stream executable not successful!!");
+        return None;
+    }
+
+    let output = Command::new("cabal")
+        .arg("list-bin")
+        .arg("program-stream")
+        .current_dir(&root)
+        .output()
+        .ok()?;
+
+    if !output.status.success() {
+        eprintln!("list-bin failed");
+        return None;
+    }
+
+    let path_str = String::from_utf8(output.stdout).ok()?;
+
+    Some(PathBuf::from(path_str.trim_end()))
+}
+
+fn stream_for_elf(elf: &Path) -> Option<Vec<u8>> {
+    let output = Command::new(&*PROG_STREAM_EXEC).arg(elf).output().ok()?;
+
+    if !output.status.success() {
+        eprintln!("ERROR");
+        return None;
+    }
+
+    Some(output.stdout)
+}
+
+lazy_static::lazy_static! {
+    static ref PROG_STREAM_EXEC: PathBuf = program_stream_exec_path().unwrap();
+}
+
+#[test]
+fn find_cabal_stuff() {
+    dbg!(&*PROG_STREAM_EXEC);
+}
+
+unsafe fn write_program_to_memory(input: &mut impl Iterator<Item = u8>) -> Option<()> {
```

Comment:
This seems to always return a trivial `Some()`. What's up with this return type? Shouldn't it return simply nothing?

Reply:
Great makes sense